src/interactive-window/interactiveWindowProvider.ts

```diff
@@ -120,40 +130,85 @@ export class InteractiveWindowProvider implements IInteractiveWindowProvider, IA
 
-        // When shutting down, we fire an event
-        const handler = result.closed(this.onInteractiveWindowClosed.bind(this, result));
-        this.disposables.push(result);
-        this.disposables.push(handler);
-        this.disposables.push(result.onDidChangeViewState(this.raiseOnDidChangeActiveInteractiveWindow.bind(this)));
+            const [inputUri, editor] = await this.createEditor(preferredController, resource, mode, commandManager);
+            const result = new InteractiveWindow(
+                this.serviceContainer.get<IDocumentManager>(IDocumentManager),
+                this.serviceContainer.get<IFileSystem>(IFileSystem),
+                this.serviceContainer.get<IConfigurationService>(IConfigurationService),
+                commandManager,
+                this.serviceContainer.get<INotebookExporter>(INotebookExporter),
+                this.serviceContainer.get<IWorkspaceService>(IWorkspaceService),
+                resource,
+                mode,
+                this.serviceContainer.get<IPythonExtensionChecker>(IPythonExtensionChecker),
+                this.serviceContainer.get<IExportDialog>(IExportDialog),
+                this.notebookControllerManager,
+                this.serviceContainer,
+                this.serviceContainer.get<IInteractiveWindowDebugger>(IInteractiveWindowDebugger),
+                this.serviceContainer.get<IDataScienceErrorHandler>(IDataScienceErrorHandler),
+                preferredController,
+                editor,
+                inputUri
+            );
+            this._windows.push(result);
 
-        // fire created event
-        this._onDidCreateInteractiveWindow.fire(result);
-        return result;
+            // This is the last interactive window at the moment (as we're about to create it)
+            this.lastActiveInteractiveWindow = result;
+
+            // When shutting down, we fire an event
+            const handler = result.closed(this.onInteractiveWindowClosed.bind(this, result));
+            this.disposables.push(result);
+            this.disposables.push(handler);
+            this.disposables.push(result.onDidChangeViewState(this.raiseOnDidChangeActiveInteractiveWindow.bind(this)));
+
+            // fire created event
+            this._onDidCreateInteractiveWindow.fire(result);
+            return result;
+        } finally {
+            creationInProgress.resolve();
+            this.pendingCreations = this.pendingCreations.filter((item) => item !== creationInProgress.promise);
+        }
+    }
+    private async createEditor(
+        preferredController: VSCodeNotebookController | undefined,
+        resource: Resource,
+        mode: InteractiveWindowMode,
+        commandManager: ICommandManager
+    ): Promise<[Uri, NotebookEditor]> {
+        const controllerId = preferredController ? `${JVSC_EXTENSION_ID}/${preferredController.id}` : undefined;
```

Comment:
If you didn't need this trace message you could do the creation of the editor at the same time as figuring out the preferred controller.

Reply:
The calling function has to wait for the preferredController and its only usage is this trace message. The calling function could instead create the editor at the same time the preferred controller is coming up.

Reply:
I moved the trace message above, however we do need the preferred controller before this step, as its used to create the IW a few lines further below.